File: Animal/Animal.java
public class Animal {
    private String name;
    private int energy;
    private int age;
    private Habitat habitat;

    public Animal(String name, int energy) {
        this.name = name;
        this.energy = energy;
        this.age = 0;
        this.habitat = null;
    }

    // Getters
    public String getName() { return name; }
    public int getEnergy() { return energy; }
    public int getAge() { return age; }
    public Habitat getHabitat() { return habitat; }
    
    // Setters
    public void setName(String name) { this.name = name; }
    public void setEnergy(int energy) { this.energy = energy; }
    public void setAge(int age) { this.age = age; }
    public void setHabitat(Habitat habitat) { this.habitat = habitat; }

    // Actions
    public void eat(int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }
        energy += amount;
    }
    
    public void growOld() {
        age++;
        if (energy >= 5) {
            energy -= 5;
        } else {
            energy = 0;
        }
    }

    // Habitat methods
    public void moveTo(Habitat newHabitat) {
        if (energy >= 10) {
            this.habitat = newHabitat;
            energy -= 10;
        }
    }
    
    public void leaveHabitat() {
        this.habitat = null;
    }
    
    public boolean hasHabitat() {
        return habitat != null;
    }
    
    public String describe() {
        if (hasHabitat()) {
            return name + " (" + energy + " energy) lives in " + habitat.getType();
        } else {
            return name + " (" + energy + " energy) is homeless";
        }
    }
    
    public int calculateDailyNeeds() {
        if (habitat == null) {
            return 20; // Basic needs
        }
        
        String type = habitat.getType();
        if (type.equals("Desert")) {
            return 40;
        } else if (type.equals("Forest")) {
            return 25;
        } else if (type.equals("Savanna")) {
            return 30;
        } else {
            return 20;
        }
    }
    
    public boolean canSurvive() {
        return energy >= calculateDailyNeeds();
    }
}

File: tp3-animal/src/animal.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Optional
from .habitat import Habitat

@dataclass
class Animal:
    name: str
    energy: int
    age: int = 0
    habitat: Optional[Habitat] = None

    # getters / setters
    def get_name(self) -> str:
        return self.name

    def get_energy(self) -> int:
        return self.energy

    def get_age(self) -> int:
        return self.age

    def get_habitat(self) -> Optional[Habitat]:
        return self.habitat

    def set_habitat(self, habitat: Optional[Habitat]) -> None:
        self.habitat = habitat

    def eat(self, amount: int) -> None:
        # Java: amount < 0 -> IllegalArgumentException("Amount must be positive")
        if amount < 0:
            raise ValueError("Amount must be positive")
        self.energy += amount

    def grow_old(self) -> None:
        # Java: age += 1
        self.age += 1
        # Java: energy -= 5
        self.energy = max(0, self.energy - 5)

    def move_to(self, habitat: Habitat) -> None:
        if self.energy >= 10:
            self.habitat = habitat
            self.energy -= 10

    def leave_habitat(self) -> None:
        self.habitat = None

    def has_habitat(self) -> bool:
        return self.habitat is not None

    def describe(self) -> str:
        if self.has_habitat():
            return f"{self.name} ({self.energy} energy) lives in {self.habitat.get_type()}"
        return f"{self.name} ({self.energy} energy) is homeless"

    def calculate_daily_needs(self) -> int:
        if self.habitat is None:
            return 20
        t = self.habitat.get_type()
        if t == "Desert":
            return 40
        if t == "Forest":
            return 25
        if t == "Savanna":
            return 30
        return 20

    def can_survive(self) -> bool:
        return self.energy >= self.calculate_daily_needs()
